### PR TITLE
Add ADR and demo code for ReactQuery

### PR DESCRIPTION
**Pull request by suzubara**
## Description

This ADR proposes using [React Query](https://react-query.tanstack.com/) to handle API data management on the frontend, starting with the new Office app pages. I also included an example, used to populate the PaymentRequestIndex page to demonstrate the differences in implementation. We can decide whether or not the code changes should be included in this PR.

### CODE REVIEW

File: src/services/ghcApi.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    // eslint-disable-next-line no-console
+    console.log('Operation does not exist', operationPath);
+  }
+
+  let request;
+  try {
+    request = operation();
+  } catch (e) {
+    // eslint-disable-next-line no-console
+    console.error('operation failed', e);
```
Comment (gidjin) [inline src/services/ghcApi.js:51]: I wonder if we should make an exception for `console.error` and maybe `console.debug`

File: src/services/ghcApi.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  return request
+    .then((response) => {
+      const payloadSchema = schema.paymentRequests;
+      return normalize(response.body, payloadSchema).entities;
+    })
+    .catch((response) => {
+      // eslint-disable-next-line no-console
+      console.log('response failed', response);
```
Comment (gidjin) [inline src/services/ghcApi.js:61]: `console.error` instead maybe?

File: src/scenes/Office/TIO/paymentRequestIndex.jsx [hunk 3]
```diff
@@ -1,52 +1,50 @@
 import React from 'react';
```
Comment (gidjin) [inline src/scenes/Office/TIO/paymentRequestIndex.jsx:1]: Should this file also move out of `src/scenes`? Though that may be more changes than this ADR should have.

File: src/scenes/Office/TIO/paymentRequestIndex.jsx [hunk 4]
```diff
@@ -1,52 +1,50 @@
...
-import { connect } from 'react-redux';
-import { bindActionCreators } from 'redux';
-import { getPaymentRequestList, selectPaymentRequests } from 'shared/Entities/modules/paymentRequests';
-import { Link, withRouter } from 'react-router-dom';
+import { Link } from 'react-router-dom';
+import { useQuery } from 'react-query';
 
-class PaymentRequestIndex extends React.Component {
-  componentDidMount() {
-    // TODO - only get payment requests associated with the given move order ID
```
Comment (gidjin) [inline src/scenes/Office/TIO/paymentRequestIndex.jsx:9]: Do you need this TODO, it didn't make it over into the new request.

File: src/services/ghcApi.js (hunk 1 above)
Comment (tinyels) [reply src/services/ghcApi.js:51]: Comment about comment: There is a STIG about not showing stack traces, so it's good eslint warns about these in general.

**Comment by suzubara** [comment]
I'm leaving the code changes in while this ADR is in review to serve as an example, but will revert them once the ADR is approved before merging.

File: src/services/ghcApi.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  let request;
+  try {
+    request = operation();
+  } catch (e) {
+    // eslint-disable-next-line no-console
+    console.error('operation failed', e);
+  }
+
+  return request
+    .then((response) => {
```
Comment (duncan-truss) [inline src/services/ghcApi.js:55]: What happens on non 2XX responses that aren't rejected, do we need access to the response code/status not just the body?

Comment (suzubara) [reply src/services/ghcApi.js:55]: yeah but this is just an example, the actual implementation would include real error handling

Comment (duncan-truss) [reply src/services/ghcApi.js:55]: Okay, I saw you mentioned ReactQuery having retry logic so thought that might be included here.  That seems to be part of the component hook?

Do we know how it integrates with the If-Match logic used by the backend?

Comment (suzubara) [reply src/services/ghcApi.js:55]: yeah there is a default retry behavior we can customize (https://react-query.tanstack.com/docs/guides/query-retries). As far as If-Match, we will need to make sure to handle this in the code that makes HTTP calls and the logic for updating the cache (which I don't believe is any different from what we do now).

**Comment by suzubara** [comment]
We will need to customize the retry behavior because [SwaggerClient](https://github.com/swagger-api/swagger-js/blob/master/docs/usage/http-client.md#status-error) actually rejects responses that are not 2xx (which differs from the `fetch` spec). So we'll probably want to modify it to only retry if the returned error is network-related (instead of on any API error).

**Comment by suzubara** [comment]
I would like to move this ADR along since it will be impactful on the impending TXO work. Any files in this PR that are _not_ the ADR itself will be reverted before merging. I think the next step after this is accepted is to work on a story that establishes some baseline patterns for using React Query with complete examples (such as tests), and ensures parity with our other API interactions. I imagine this would include:

- Identify an existing TXO API interaction that can be swapped out for React Query and used as a first example
- Add a function that mirrors `SwaggerRequest` behavior to make actual API calls (but outside of the context of Redux)
- Ensure configuration (such as retries, error logging, etc.) meets our specific needs (documentation [here](https://react-query.tanstack.com/docs/api#reactqueryconfigprovider))
- Include tests and any utilities needed to test components that use React Query hooks (ie, [ReactQueryCacheProvider](https://react-query.tanstack.com/docs/api#reactquerycacheprovider))

If there are any other specific requirements or concerns about the above, please let me know!

**Pull request merged**